bundles/io/org.eclipse.smarthome.io.voice/src/main/java/org/eclipse/smarthome/io/voice/text/AbstractRuleBasedInterpreter.java

```diff
@@ -106,23 +107,89 @@ public String interpret(Locale locale, String text) throws InterpretationExcepti
     }
 
     private void invalidate() {
-        identifierTokens = null;
+        allItemTokens = null;
+        itemTokens = null;
         languageRules = null;
     }
 
     /**
      * All the tokens (name parts) of the names of all the items in the {@link ItemRegistry}.
      *
+     * @param locale The locale that is to be used for preparing the tokens.
      * @return the identifier tokens
      */
-    HashSet<String> getIdentifierTokens() {
-        if (identifierTokens == null) {
-            identifierTokens = new HashSet<String>();
+    HashSet<String> getAllItemTokens(Locale locale) {
+        if (allItemTokens == null) {
+            allItemTokens = new HashMap<Locale, HashSet<String>>();
+        }
+        HashSet<String> localeTokens = allItemTokens.get(locale);
+        if (localeTokens == null) {
+            allItemTokens.put(locale, localeTokens = new HashSet<String>());
             for (Item item : itemRegistry.getAll()) {
-                identifierTokens.addAll(splitName(item.getLabel(), true));
+                localeTokens.addAll(tokenize(locale, item.getLabel()));
+            }
+        }
+        return localeTokens;
+    }
+
+    /**
+     * Retrieves the list of identifier token sets per item currently contained in the {@link ItemRegistry}.
+     *
+     * @param locale The locale that is to be used for preparing the tokens.
+     * @return the list of identifier token sets per item
+     */
+    HashMap<Item, ArrayList<HashSet<String>>> getItemTokens(Locale locale) {
+        if (itemTokens == null) {
+            itemTokens = new HashMap<Locale, HashMap<Item, ArrayList<HashSet<String>>>>();
+        }
+        HashMap<Item, ArrayList<HashSet<String>>> localeTokens = itemTokens.get(locale);
+        if (localeTokens == null) {
+            itemTokens.put(locale, localeTokens = new HashMap<Item, ArrayList<HashSet<String>>>());
+            for (Item item : itemRegistry.getItems()) {
+                if (item.getGroupNames().isEmpty()) {
+                    addItem(locale, localeTokens, new HashSet<String>(), item);
+                }
+            }
+        }
+        return localeTokens;
+    }
+
+    private void addItem(Locale locale, HashMap<Item, ArrayList<HashSet<String>>> target, HashSet<String> tokens,
+            Item item) {
+        HashSet<String> nt = new HashSet<String>(tokens);
+        nt.addAll(tokenize(locale, item.getLabel()));
+        ArrayList<HashSet<String>> list = target.get(item);
+        if (list == null) {
+            target.put(item, list = new ArrayList<HashSet<String>>());
+        }
+        list.add(nt);
+        if (item instanceof GroupItem) {
+            for (Item member : ((GroupItem) item).getMembers()) {
+                addItem(locale, target, nt, member);
+            }
+        }
```

Comment:
Now I'm more confused. Each item has its own list, and when you call addItem recursively, you're passing a different Item object (assuming that items are not members of themselves, in which case, you've got an infinite recursion issue). So it looks to me as if each list will always have exactly one element. So I don't see why the data structure needs the ArrayList at all...   Why isn't it just a Map<Item, Set<String>>?  Maybe there is code below that I haven't read yet that will make this clear...

Separately, if I'm reading this correctly, the tokens for each item will include the tokens for any groups it is part of as well. That seems like it is worth documenting with a comment here or with something in the doc comment on getItemTokens()